### ISSUE TITLE
Consider preventing setting finidat in a hybrid case

### ISSUE THREAD

**Issue by billsacks** [ESCOMP/CTSM#1082]
A large production run was recently discovered to have a problem: This run was broken into 100-year segments with a separate case for each segment, using hybrid starts from the previous segment. However, the user running the cases accidentally left a setting in `user_nl_clm` from the first case: setting finidat to point to the very beginning initial conditions. Thus, every 100 years, CTSM was resetting its conditions back to the beginning. Unfortunately, this error wasn't discovered until the run was complete.

@ekluzek suggested putting in place a warning if build-namelist sees a setting of `finidat` in `user_nl_clm` in a hybrid case. He suggests using the mechanism where this will cause build-namelist to die (i.e., treating warnings as errors) unless the `ignore_warnings` flag is set. This makes sense to me.

However, before we do this, we'd like to know how often users legitimately want to override `finidat` in a hybrid case. @olyson - we're wondering if you have a sense of this?

I know that there are some other flags that would cause similar issues if they are accidentally kept in place in subsequent cases of a linked set of cases like the one described above. For example, the various `reset_snow` namelist flags come to mind; however, I think it's likely that users *will* want to set the `reset_snow` flags in a hybrid case, so my gut feeling is that we don't want to apply this check to those flags. I'm not sure about other flags; for now, it's probably good enough to target the single `finidat` flag, since that is likely to be the most common source of errors.

**Comment by olyson** [comment]
Just so I'm clear on this, a hybrid run will normally set finidat to a restart file specified by the RUN_REFCASE and RUN_REFDATE, right?
But if one was sequentially cloning cases starting from a case that had finidat set in user_nl_clm and it was not removed from the user_nl_clm in the new cases, then the model doesn't complain about that, even if the REFCASE and REFDATE don't match?
I don't know of situations where one would want to override finidat in a hybrid case, unless there are cases related to data assimilation or short-term prediction that I'm not aware of where one would want to do that.
May I ask what production run is affected?

**Comment by billsacks** [comment]
Thanks, @olyson - your restatement of the problem is exactly right.

The affected run was one done by the paleo & land ice working groups: a Last Interglacial run.

**Comment by klindsay28** [comment]
A usecase for setting finidat in user_nl_clm in a hybrid run is mixing IC files for different components from different ref cases. An example that comes to mind is initializing the BGC in the coupled model using spun up BGC state from CLM/CTSM and POP from I and G compset experiments respectively.

When I encounter this usecase, I usually set RUN_REFCASE and RUN_REFDATE for the ref case common to most components and rely on user_nl for setting the IC in the other components.

**Comment by ekluzek** [comment]
@klindsay28 thanks for pointing that out. For the case of CLM ,at least for releases, we've typically put together a REFCASE that includes these mixed components. So the user wouldn't have to do this themselves. But, that's only for really standard things that we want people to easily replicate.

Our current proposed fix for this isn't to totally disallow this type of usage -- it's to have it fail the first time -- and make the user add an "ignore-warnings" flag that says they know what they are doing. So your message tells me that that's the right approach rather than just never allowing this at all.

**Comment by klindsay28** [comment]
Make sense @ekluzek.

That said, if this mechanism was in place for the last interglacial run, would it have helped? If the person that ran the last interglacial run needed to set finidat in user_nl_clm for their first 100-year segment, wouldn't they also have needed to set ignore-warnings in that case? If so, it seems like they might have just propagated this ignore-warning setting to subsequent cases, as they did for finidat, and no error would have been triggered in the subsequent cases.

**Comment by billsacks** [comment]
@klindsay28 you raise a very good point. I just checked, and indeed the first run in the series also appears to have been a hybrid case. We should get some feedback from the affected user and others to see if something like this would have helped.

**Comment by billsacks** [comment]
Actually, I don't think it's even necessary for me to check in with the original user on @klindsay28 's point, because I see that the series of cases were indeed created with `create_clone`. The way I could see around this is:

- Make the `ignore_warnings` flag an xml variable

- Add some logic to `create_clone` that lets it reset select xml variables to their default value in a clone. I'm not sure how hard this will be to do in cime, but will open an issue to gather thoughts on it

**Comment by ekluzek** [comment]
Good point. I do want to say that there is one advantage of the "ignore_warnings" setting in that it will print out about it's use when you run preview_namelists. So even though your case will be setup this way from the "create_clone" it will print that you are ignoring warnings and you got this list of specific warnings. So it is better -- but not perfect. Bill's mechanism would be required to make this still more robust.

**Comment by ekluzek** [comment]
The other option would be to make it a fatal error if finidat is set for a hybrid case. You could still get around it by setting up a REFCASE directory by copying in the finidat file that you want to. use. Or you could comment out the fatal error in the build-namlist. So you could still do the thing that @klindsay28 points out, with more work. I'm kind of wondering if that isn't the most robust solution for this.

@wwieder, @dlawrenncar, and @danicalombardozzi can you comment on how important you think this is to fix? My inclination is that because it screwed up both a set of simulations and a scientific paper on them -- that it's pretty important to fix. In looking at this we also realize that it's a pretty subtle error to notice. So it could easily be a problem that others do and even that others have done without it being caught. As such I could put at least one of our suggested fixes in one of the upcoming bit-for-bit tags that are already planned.

**Comment by billsacks** [comment]
I don't think we should jump into changing this immediately, but instead should wait until we have a solution that we're happy with: In my mind, the fact that it caused a major problem for a single case means that it's worth thinking about whether there's a way to fix it, but that doesn't make it a critical, high-priority issue, since this behavior has been in place for a very long time. I'm also wary about adding a mechanism that creates too much inconvenience for legitimate use cases: I feel that making this a hard error may be going too far. I think it's also worth discussing this with CSEG, because ideally we'd have some consistency between components in this respect.

Also: Assuming we go the route of making this a "soft" (bypassable) error, I'm starting to think that it would be best to have a dedicated xml variable for this purpose rather than using a blanket "ignore warnings" variable. I think it would be clearest and least error prone to have an xml variable like `ALLOW_HYBRID_OVERRIDE_LND` which would need to be set to TRUE in order to override the hybrid refcase with finidat. (Note that my naming convention here assumes that we might want to have similar variables for the other components... thinking about consistency between components here.)

**Comment by dlawrenncar** [comment]
I agree that fixing this is not urgent, but also that a fix is
likely required to avoid future mistakes.  I think this is best discussed
at CSEG level to agree on a common method across components.

**Comment by billsacks** [comment]
I talked a bit with @mvertens about this. She feels like it might be best just to make this a hard error, and require that users create a new refcase if they want to mix and match initial conditions like this. I see her point that it can be better to just have one way to do something, rather than having multiple ways, one of which requires setting multiple flags. One downside of this approach is that it is currently difficult to create a refcase, because you need to rename all of the files to have the case name of the refcase. So actually, as I write this, I am seeing a problem with this suggestion, in that it makes the case setup less self-documenting, particularly since the provenance of your custom refcase won't be captured in any of the new case's metadata. So now I'm back to being on the fence about this myself.

We'll discuss this at tomorrow's CSEG meeting. @klindsay28 do you have feelings on this - i.e., strong preferences between needing to create a new refcase or continuing to use your current workflow but setting some additional flag to allow overriding initial conditions?

**Comment by klindsay28** [comment]
I personally do not like the idea of forcing a user to create a new refcase to be able to combine model states from different experiments. In my opinion, this usecase is one of the primary purposes for having RUN_TYPE=hybrid.

I would strongly object if I couldn't specify IC in user_nl_pop in hybrid runs. I regularly specify BGC IC in user_nl_pop, separate from IC for physical state. I would strongly object to adding complexity to this common use case.

I initially thought I would write that this should be a CTSM developer/user community decision to impose such a constraint (or not) for CTSM only and that I would only strongly object only if this were imposed on specification in user_nl_pop. However, it is not uncommon for CESM users to complain about non-uniformity across components within CESM. I think this is a legit point of view and decisions that introduce differences between components should consider this point of view (acknowledging that you might move forward with such a change after such a consideration).

Another point to consider is what amount of vetting of a case is the responsibility of the user. One (extreme) point of view is that if the user puts a nml var setting into user_nl_*, then they know what they're doing and the model should do what the user tells them to do. Taken fully, this point of view is probably too extreme, and having the model do some sanity/safety checks on the inputs makes sense. That said, I think some of the responsibility lies with the user, instead of increasing the burden for other users.

**Comment by billsacks** [comment]
@klindsay28 thank you for sharing your thoughts. I really appreciate it. I do feel this is something that should be done consistently across components: while this particular case was messed up by a CTSM `finidat` setting, it seems just as likely that a case would be messed up in the future by a similar setting for (say) the ocean. I feel that doing things differently between components could do more harm than good, for example if a user comes to rely on some error checking that is done for one component, not realizing that it isn't done for another.

Can you summarize your feeling at this point regarding forcing a user (like yourself) to make an additional xmlchange setting in the situation where you want to override one of the component's initial condition files in a hybrid case? So you could still do it, but would need to add something like `./xmlchange ALLOW_HYBRID_OVERRIDE=TRUE` to indicate that you know what you're doing. Would that be acceptable?

**Comment by klindsay28** [comment]
I would be fine with needing to make an additional xmlchange to override IC files in a hybrid. My strong objection was to make this a hard error and require the creation of a new refcase.

I do worry that we're back to not solving the last interglacial run's problem, that the ALLOW_HYBRID_OVERRIDE=TRUE setting would get propagated via create_clone. Perhaps the variable's name would have caused them to pause enough to realize that they don't want that setting in subsequent cases. It's hard to know what the user would have done.

I'm curious how the CSEG discussion will go. Here's a tidbit of info for that conversation. I took a quick look at cases in /glade/p/cesmdata/cseg/runs/cesm2_0, where CESM2 development runs were placed. I count 530 b runs with RUN_TYPE=hybrid, and 200 of these have finidat set in user_nl_clm. I think most of these runs were done by Cecile (I didn't count how many). Her input on this topic might be useful.

**Comment by billsacks** [comment]
@klindsay28 - Thanks a lot for your additional thoughts. Your very good point about propagating `ALLOW_HYBRID_OVERRIDE=TRUE` prompted me to open https://github.com/ESMCI/cime/issues/3621. I haven't looked closely, but my sense is that this could be implemented without too much trouble (famous last words, I know). Admittedly, though, it would add to the complexity of the system, so I truly don't know what the right answer is here.

I'll also ping Cecile on this.

**Comment by billsacks** [comment]
In response to these three options:

> (1) Maintain the status quo, figuring that this was a one-time unfortunate occurrence, but not worth trying to address
> 
> (2) "Hard fail": Not allow overriding the hybrid refcase's initial conditions at all – instead, requiring that a new refcase be created for this use case
> 
> (3) Intermediate solution: "Soft fail": Overriding a refcase's initial conditions for any component results in an error unless you additionally set the xml variable ALLOW_HYBRID_OVERRIDE=TRUE. This xml variable would be reset to FALSE when running create_clone, forcing you to again specify that you really mean to do this.

@cecilehannay says (by email):

> I agree with Keith that option (2) is not the way to go. I like option (3)

**Comment by billsacks** [comment]
Based on yesterday's CSEG meeting, we are not going to do anything about this for now.

There were mixed feelings about this expressed at the CSEG meeting:

- Some feelings are that users should check things like this carefully

- On the other hand, others (and some of the same people) feel that the system is so complex that there's a lot of burden on users

- On the third hand, this could give a false sense of security, because this probably wouldn't check everything that should be checked in this case

The overall consensus, though, is that this isn't worth trying to address.
